packages/idempotency/tests/e2e/idempotencyDecorator.test.FunctionCode.ts

```diff
@@ -0,0 +1,95 @@
+import { LambdaInterface } from '@aws-lambda-powertools/commons';
+import { DynamoDBPersistenceLayer } from '../../src/persistence';
+import { idempotent } from '../../src/idempotentDecorator';
+import { Context } from 'aws-lambda';
+import { Logger } from '../../../logger';
+
+const IDEMPOTENCY_TABLE_NAME = process.env.IDEMPOTENCY_TABLE_NAME;
+const dynamoDBPersistenceLayer = new DynamoDBPersistenceLayer({
+  tableName: IDEMPOTENCY_TABLE_NAME,
+});
+
+const ddbPersistenceLayerCustomized = new DynamoDBPersistenceLayer({
+  tableName: IDEMPOTENCY_TABLE_NAME,
+  dataAttr: 'dataattr',
+  keyAttr: 'customId',
+  expiryAttr: 'expiryattr',
+  statusAttr: 'statusattr',
+  inProgressExpiryAttr: 'inprogressexpiryattr',
+  staticPkValue: 'staticpkvalue',
+  validationKeyAttr: 'validationkeyattr',
+});
+
+interface TestEvent {
+  [key: string]: string
+}
+
+interface EventRecords {
+  records: Record<string, unknown>[]
+}
+
+class DefaultLambda implements LambdaInterface {
+
+  @idempotent({ persistenceStore: dynamoDBPersistenceLayer })
+  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+  // @ts-ignore
+  public async handler(_event: Record, _context: Context): Promise<string> {
+    logger.info(`Got test event: ${JSON.stringify(_event)}`);
+    // sleep to enforce error with parallel execution
+    await new Promise((resolve) => setTimeout(resolve, 3000));
+
+    return 'Hello World';
+  }
+
+  @idempotent({ persistenceStore: ddbPersistenceLayerCustomized })
+  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+  // @ts-ignore
+  public async handlerCustomized(_event: TestEvent, _context: Context): Promise<string> {
+    logger.info(`Got test event customized: ${JSON.stringify(_event)}`);
+    // sleep for 5 seconds
+
+    return 'Hello World Customized';
+  }
+
+  @idempotent({ persistenceStore: dynamoDBPersistenceLayer })
+  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+  // @ts-ignore
+  public async handlerFails(_event: TestEvent, _context: Context): Promise<string> {
+    logger.info(`Got test event: ${JSON.stringify(_event)}`);
+    // sleep for 5 seconds
+
+    throw new Error('Failed');
+  }
+
+}
+
+const defaultLambda = new DefaultLambda();
+export const handler = defaultLambda.handler.bind(defaultLambda);
+export const handlerCustomized = defaultLambda.handlerCustomized.bind(defaultLambda);
+export const handlerFails = defaultLambda.handlerFails.bind(defaultLambda);
+
+const logger = new Logger();
+
+class LambdaWithKeywordArgument implements LambdaInterface {
+  public async handler(_event: EventRecords, _context: Context): Promise<string> {
+    logger.info(`Got test event: ${JSON.stringify(_event)}`);
+    for (const record of _event.records) {
+      logger.info(`Processing event: ${JSON.stringify(record)}`);
+      await this.process(record);
+    }
+
+    return 'Hello World Keyword Argument';
+  }
+
+  @idempotent({ persistenceStore: dynamoDBPersistenceLayer, dataKeywordArgument: 'foo' })
```

Comment:
I agree to introduce dedicated decorator for the handler to make the input more explicit for each case. Let's have `idempotentLambdaHandler` and `idempotentMethod` similar to the Tracer. 

> As a matter of fact, the dataKeywordArgument (at least the way I understand it in Python) refers to the name of the argument of the function that contains the data to be used for the idempotency, in this case it'd be record. Not the key (i.e. foo) in the payload.

This was [discussed in the RFC previously](https://github.com/awslabs/aws-lambda-powertools-typescript/issues/447#issuecomment-1348681945). The agreement was enforce the signature of the function wrapper to a record and use `datakeywordArgument` to decide which key/value to use for idempotency.

Reply:
Okay, I'm still a bit on the fence about this and not 100% sure about the agreement that was reached in the RFC now that I understand better the Idempotency utility.

I propose this, let's put a pin on the `dataKeywordArgument` topic for now and follow the RFC, so that we don't block the development. When we are done with all the features let's try to use it in real codebases and see how the DX is, as well as hearing customer feedback. If needed we'll change it then.